### ISSUE TITLE
`ml-cuda10.1-4.1.0.json` generate `cuda11.1` definition files

### ISSUE THREAD

**Issue by eitsupi** [rocker-org/rocker-versioned2#182]
Unlike the file name, `ml-cuda10.1-4.1.0.json` generates Dockerfile and compose file of cuda11.1.

Because a Dockerfiles such as `Dockerfile_r-ver_4.1.0-cuda11.1` are generated by `ml-cuda11.1-4.1.0.json` later and overwrites it, the image built by GitHubActions to build the image of cuda10.1 is actually the image defined in `ml-cuda11.1-4.1.0.json`.
The contents of both are quite different.

This doesn't seem to be an intentional behavior, do you think it can be solved by auto-generating stack files ?

```shell
$ diff ./stacks/ml-cuda10.1-4.1.0.json ./stacks/ml-cuda11.1-4.1.0.json
5c5
<   "LABEL": "org.label-schema.license=\"GPL-2.0\" \\\n      org.label-schema.vcs-url=\"https://github.com/rocker-org/rocker-versioned\" \\\n      org.label-schema.vendor=\"Rocker Project\" \\\n      maintainer=\"Carl Boettiger <cboettig@ropensci.org>\"",
---
>   "LABEL": "org.opencontainers.image.licenses=\"GPL-2.0-or-later\" \\\n      org.opencontainers.image.source=\"https://github.com/rocker-org/rocker-versioned2\" \\\n      org.opencontainers.image.vendor=\"Rocker Project\" \\\n      org.opencontainers.image.authors=\"Carl Boettiger <cboettig@ropensci.org>\"",
9c9,10
<     "FROM": "rocker/r-ver:4.1.0",
---
>     "FROM": "nvidia/cuda:11.1.1-cudnn8-devel-ubuntu20.04",
>     "COPY": "scripts /rocker_scripts",
11,19c12,18
<       "CUDA_VERSION": "11.1",
<       "NCCL_VERSION": "2.7.8",
<       "NVIDIA_VISIBLE_DEVICES": "all",
<       "NVIDIA_DRIVER_CAPABILITIES": "compute,utility",
<       "NVIDIA_REQUIRE_CUDA": "cuda>=11.1 brand=tesla,driver>=418,driver<419 brand=tesla,driver>=440,driver<441 brand=tesla,driver>=450,driver<451",
< 
<       "CUDA_HOME": "/usr/local/cuda",
<       "LD_LIBRARY_PATH": "$LD_LIBRARY_PATH:$CUDA_HOME/lib64:$CUDA_HOME/extras/CUPTI/lib64:$CUDA_HOME/lib64/libnvblas.so:",
<       "LIBRARY_PATH": "/usr/local/cuda/lib64/stubs",
---
>       "R_VERSION": "4.1.0",
>       "TERM": "xterm",
>       "LC_ALL": "en_US.UTF-8",
>       "LANG": "en_US.UTF-8",
>       "R_HOME": "/usr/local/lib/R",
>       "CRAN": "https://packagemanager.rstudio.com/all/__linux__/focal/latest",
>       "TZ": "Etc/UTC",
24,25c23,24
<       "RETICULATE_MINICONDA_ENABLED": "FALSE",
<       "PATH": "${PYTHON_VENV_PATH}/bin:${CUDA_HOME}/bin:/usr/local/nviida/bin:${PATH}:/usr/local/texlive/bin/x86_64-linux"
---
>       "RETICULATE_AUTOCONFIGURE": "0",
>       "PATH": "${PYTHON_VENV_PATH}/bin:${PATH}:${CUDA_HOME}/bin"
28c27
<       "/rocker_scripts/install_cuda-11.1.sh",
---
>       "/rocker_scripts/install_R.sh",
37,41c36,37
<       "S6_VERSION": "v1.21.7.0",
<       "RSTUDIO_VERSION": "latest",
<       "PANDOC_VERSION": "default",
<       "TENSORFLOW_VERSION": "gpu",
<       "KERAS_VERSION": "default",
---
>       "S6_VERSION": "v2.1.0.2",
>       "RSTUDIO_VERSION": "preview",
61c57
<   }  
---
>   }
```

The compose files generated by both of them is exactly the same.
https://github.com/rocker-org/rocker-versioned2/blob/e263a0a456d638a01bceaa225b953fb3041ca22d/compose/ml-cuda10.1-4.1.0.yml
https://github.com/rocker-org/rocker-versioned2/blob/e263a0a456d638a01bceaa225b953fb3041ca22d/compose/ml-cuda11.1-4.1.0.yml

```shell
$ diff ./compose/ml-cuda10.1-4.1.0.yml ./compose/ml-cuda11.1-4.1.0.yml

```

**Comment by cboettig** [comment]
Thanks for catching that!  yup, definitely looks like a typo on the stack cuda10.1 stack file which would be avoided by auto-generating.  

Pinning the CUDA versions is still somewhat new territory for us.  Extrapolating from the version policy used elsewhere would suggest that `latest` should use the most recent available version, and then become pinned to that.  (i.e. probably 11.3.1 right now... https://hub.docker.com/r/nvidia/cuda/tags?page=1&ordering=last_updated).  I'm not entirely sure if that's the best policy here though, GPU lib development moves at a breakneck pace these days, and unlike any other version software we package, the cuda toolkit version should not be any newer than than the _driver version_ of the host machine.  For example, a host running the current Ubuntu LTS will have cuda driver version 10.1 by default, and so ought not run our images with cuda toolkit 11.1,

- https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions, 
- https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#cuda-runtime-and-driver-api-version

We might also look more closely at how other projects are handling this, e.g. official [tensorflow gpu](https://hub.docker.com/r/tensorflow/tensorflow/tags?page=1&ordering=last_updated&name=gpu) images.  Looks like the current 2.4.2 version is pinned to CUDA 11.0.3 , and devel to 11.2.1, though not sure what the logic rule is.  (current `nvidia/cuda` is [11.3.1](https://hub.docker.com/r/nvidia/cuda/tags?page=1&ordering=last_updated)).  

So any ideas for how best to choose what version(s) we want to support would be most welcome!  


Anyway, thanks for catching the error, and automated stack files with less redundancy would certainly help avoid that!

**Comment by eitsupi** [comment]
Thank you for the explanation.
I understand almost nothing about cuda. I'm also confused by the fact that the cuda docker image exists on nvcr.io as well as on DockerHub.

Anyway, I looked at `ml-cuda10.1-x.x.x.json` and `ml-cuda11.1-x.x.x.json` again, and it seems that CUDA version rarely changes, so I thought we could use the same method as #169 and #175 that copy `devel` to the latest two versions' definition automatically.

Since ml-cuda10.1 can only be built after `rocker/r-ver`, it would be better to organize the workflow in `core.yml` as I did in #175, or change the trigger to work after `core.yml` is completed.
https://docs.github.com/actions/reference/events-that-trigger-workflows#workflow_run

Also, since these images install RStudio by the same procedure as `rocker/rstudio`, it may be useful to copy `S6_VERSION` of `core-devel.json` to `ml-cuda10.1-devel.json` and `ml-cuda11.1-devel automatically.
(Since it is getting more and more complicated, it would be better to create a directory like `build/` and move the scripts into it, and write in `build/README.md` which files are to be edited manually and which files are to be generated automatically?)
